imgui/src/widget/text_filter.rs

```diff
@@ -0,0 +1,104 @@
+use crate::sys;
+use crate::Ui;
+use std::ptr;
+
+/// Helper to parse and apply text filters
+pub struct TextFilter {
+    id: String,
+    raw: *mut sys::ImGuiTextFilter,
+}
+
+impl TextFilter {
+    /// Creates a new TextFilter with an empty filter.
+    ///
+    /// This is equivalent of [new_with_filter](Self::new_with_filter) with `filter` set to `""`.
+    pub fn new(label: String) -> Self {
+        Self::new_with_filter(label, String::new())
+    }
+
+    /// Creates a new TextFilter with a custom filter.
+    pub fn new_with_filter(label: String, mut filter: String) -> Self {
+        filter.push('\0');
+        let ptr = filter.as_mut_ptr();
+        Self {
+            id: label,
+            raw: unsafe { sys::ImGuiTextFilter_ImGuiTextFilter(ptr as *mut sys::cty::c_char) },
+        }
+    }
+
+    /// Builds the TextFilter with its filter attribute. You can use
+    /// [`pass_filter()`](Self::pass_filter) after it.
+    ///
+    /// If you want control the filter with an InputText, check [`draw()`](Self::draw).
+    pub fn build(&self) {
+        unsafe {
+            sys::ImGuiTextFilter_Build(self.raw);
+        }
+    }
+
+    /// Draws an [InputText](crate::input_widget::InputText) to control the filter of the TextFilter.
+    ///
+    /// This is equivalent of [draw_with_size](Self::draw_with_size) with `size` set to `0.0`.
+    pub fn draw(&self) {
+        self.draw_with_size(0.0);
+    }
+
+    /// Draws an [InputText](crate::input_widget::InputText) to control the filter of the TextFilter.
+    ///
+    /// The InputText has the size passed in parameters.
+    pub fn draw_with_size(&self, size: f32) {
+        unsafe {
+            let mut id = self.id.clone();
+            id.push('\0');
+            let ptr = id.as_mut_ptr();
+            sys::ImGuiTextFilter_Draw(self.raw, ptr as *mut sys::cty::c_char, size);
+        }
+    }
+
+    /// Returns true if the filter is not empty (`""`).
+    pub fn is_active(&self) -> bool {
+        unsafe { sys::ImGuiTextFilter_IsActive(self.raw) }
+    }
+
+    /// Returns true if the buffer matches the filter.
+    ///
+    /// [`draw()`](Self::draw) or [`build()`](Self::build) mut be called **before** this function.
+    pub fn pass_filter(&self, mut buf: String) -> bool {
```

Comment:
Might be worth having this take `&str` instead, as otherwise it forces on the user to do something like

```rust
State {
   options: Vec<String>,
   filter: TextFilter::new(...),
}

fn render(ui: &imgui::Ui) {
    for x in &self.options {
        // currently would need:
        filter.pass_filter(x.clone());
    }
}
```

By taking `&str` I think it better describes what the argument actually is (a read-only string used to check if it matches)

It would mean the `pass_filter` method internally allocates a new String each time, but, we can probably optimize this away with the `imgui::string::UiBuffer` later (without changing the API)

Reply:
Oh you're right. I just must create a String to add the `\0` at the end. I'll check `imgui::string::UiBuffer` too